### ISSUE TITLE
Add filter to WaitForValueChangedRequest

### ISSUE THREAD

**Issue by ghost** [NordicSemiconductor/Android-BLE-Library#63]
With `WaitForValueChangedRequest`, we can write on a characteristic and wait for the response

```kotlin
waitForNotification(responseCharacteristic)
                    .trigger(writeCharacteristic(gattCharacteristicWrite, payload))
                    .awaitValid(response)
```

The issue is that it simply returns the first value notified on `responseCharacteristic`, and there's no way to set a filter

In our case, there can be multiple writes on the same characteristic, but the response will always contain the commandId at byte[0]. So, this'd be my feature request

```kotlin
waitForNotification(responseCharacteristic)
                    .trigger(writeCharacteristic(gattCharacteristicWrite, payload))
                    .filter(response -> response[0] == expectedCommandId)
                    .awaitValid(response)
```

`filter` could take a `Predicate<Boolean>`, for example

Furthermore, there could be multiple operations enqueued writing the same commandId, and they should deliver the results in a FIFO way

Could this be a problematic scenario?
1. Read name -> "old"
2. Set name -> "new"
3. Read name -> ??

Depending on the state of the queue, could the 3rd operation receive either "old" or "new"?

I'm not sure if it'd be problematic or not, considering there's a single queue, maybe it'd always be ok.

**Comment by philips77** [comment]
Did you try using [`merge(DataMerger)`](https://github.com/NordicSemiconductor/Android-BLE-Library/blob/master/ble/src/main/java/no/nordicsemi/android/ble/WaitForValueChangedRequest.java#L124) for purpose of filtering?

https://github.com/NordicSemiconductor/Android-BLE-Library/blob/09e7adb6a93354cf07392285572a3d072dadf989/ble/src/main/java/no/nordicsemi/android/ble/data/DataMerger.java#L31-L41

From this method you could return true only if the right message was received:

```java
boolean merge(@NonNull final DataStream output,
			  @Nullable final byte[] lastPacket, @IntRange(from = 0) final int index) {
   if (lastPacket != null && lastPacket.length > 0 && lastPacket[0] == SOME_COMMAND_ID) {
      stream.write(lastPacket);
      return true;
   }
   return false;
}

It wasn't designed for this purpose, but I guess it could be used here. Of course your filtering method would be nicer.
```

**Comment by maragues-kolibree** [comment]
That seems to work 👏 

Would it make sense to create a PR adding a `filter` that does this under the hood?

**Comment by philips77** [comment]
You should be able to set filter AND the merger (the merger would merge only those packets that match filter). So implementing filter in the library must use some additional code, not `merge`. But the code could look very alike. 
What would your vision of the API using `Predicate<Boolean>`?

**Comment by maragues-kolibree** [comment]
This is how I would like to use the API, if that was your question (c&p from first message)

```waitForNotification(responseCharacteristic)
                    .trigger(writeCharacteristic(gattCharacteristicWrite, payload))
                    .filter(response -> response[0] == expectedCommandId)
                    .awaitValid(response)```

I'm not sure if this was your question.

I'm only thinking about `WaitForValueChangedRequest`, I don't know if it makes sense in other requests

**Comment by Maragues** [comment]
I'll test the fix when version 2.1 is released.

We recently migrated to `includeBuild` and I'm having issues referencing a project in my filesystem, sorry

**Comment by philips77** [comment]
Released in 2.1.0.
